### ISSUE TITLE
mingw without MSYS/MSYS2 for openssl 1.1.1

### ISSUE THREAD

**Issue by misery** [openssl/openssl#6111]
Hi there,

since OpenSSL 1.1.0 the build system checks for a new perl version and line breaks which makes MSYS1 obsolete. I appreciate this but I would like to avoid MSYS2, too. Since the build system of OpenSSL is unified for MSVC builds, too. I wonder why I still need MSYS2 as we do not need it for MSVC builds - which is great!

Most people uses this mingw project to build the software. There is a "mingw32-make" included which works perfect without any MSYS/MSYS2 stuff.
https://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win32/Personal%20Builds/mingw-builds/7.3.0/threads-posix/dwarf/

The MSVC build of OpenSSL can use ActiveState Perl without problems. So is it possible to use that Perl with MinGW only? Without MSYS2 like MSVC build.
Qt5 is also able to build with MinGW (mingw32-make) only and without MSYS.

**Comment by misery** [comment]
Of course... this is the incorrect way. I removed the line break check to see what happens. Seems he uses shell syntax. Is it possible to switch to another syntax for mingw-only builds? How MSVC makefile generation works here?

```
mingw32-make[1]: Entering directory 'C:/J/workspace/default_Review_Libs_Win32_GNU/build/b/src/openssl'
CC="gcc " C:\Perl64\bin\perl.exe crypto\aes\asm\aes-586.pl coff -m32 -Wall -O3 -fomit-frame-pointer -fstack-protector-strong -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_BN_ASM_PART_WORDS -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DRC4_ASM -DMD5_ASM -DRMD160_ASM -DAES_ASM -DVPAES_ASM -DWHIRLPOOL_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DPADLOCK_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"C:\J\workspace\default_Review_Libs_Win32_GNU\build\dist\ssl\"" -DENGINESDIR="\"C:/J/workspace/default_Review_Libs_Win32_GNU/build/dist/lib/engines-1_1\"" -DUNICODE -D_UNICODE -DWIN32_LEAN_AND_MEAN -D_MT -DOPENSSL_USE_APPLINK -DNDEBUG   crypto\aes\aes-586.s
'CC' is not recognized as an internal or external command,
operable program or batch file.
Makefile:745: recipe for target 'crypto\aes\aes-586.s' failed
```

**Comment by levitte** [comment]
Unfortunately, our mingw config targets assume a POSIX shell, so you're a bit out of luck.  A mingw build in Windows command prompt will demand some work, as far as I can see, and we're much too late in the 1.1.1 release cycle to consider it now.  I'll label this as a feature request for post 1.1.1

**Comment by dot-asm** [comment]
mingw32-make appears to be quite a slippery thing... Consider following makefile
```
all:
    FOO=bar env
```
If mingw32-make is invoked in MSYS2 window, it works. But if invoked in CMD window, it doesn't (failing with similar error message as reported). So it kind of thrives on shell that invoked it. Question would be which one do we choose and how do we explain it to user. I understand that suggestion is use it with CMD, but it would still need to be explained...

And it looks like it would have to be separately maintained. What does this mean? Great thing about "binding" mingw build procedure to MSYS2 is that it comes for free. We just reuse everything from Unix, reuse to degree that we are pretty confident that whatever we do on Unix will just work on MSYS2. Now one can wonder if it would be possible to enjoy same kind of convenience if we go for mingw32-make under CMD? Or in other words would it be possible to re-use windows-makefile.tmpl? Well, consider following makefile:
```
all:
    set FOO=bar
    set
```
mingw32-make invokes cmd *twice* and as you can imagine FOO won't be observed in set's output. nmake invokes cmd once, however, it's not like it passes both lines to cmd, instead it apparently interprets first line internally, and then invokes 'cmd /c set'. Who would have thought... Not to mention that windows-makefile.tmpl is too "invested" into MSVC even otherwise, most notably the way long lists are passed... Bottom line is that chances are that mingw32-make would have to be served by own template, which is actually undesired, because it would require explicit attention and maintenance. Alternatively it would take additional non-trivial surgery to the way makefile is generated (recall discussion about toolchain-specific rules).

**Comment by misery** [comment]
Yeah, ok, I understand that it would be another kind of build variant that needs to maintained. But it feels a little bit stepmotherly to support a major OS like Windows with a "useless" compatibility layer of Linux. :-)

But what do you think to unify MSVC, mingw with MSYS and mingw without MSYS? I think that would be the best to have one for all Windows build variants. I think you know ninja-build. Maybe it is a better idea to replace *-makefile.tmpl by a Configure generated ninja build template? That will support mingw with and without MSYS and also it supports MSVC compiler and avoids this crappy nmake.

Would that be possible with current Configure-Perl-Script?

By the way.... that could be used on Linux and FreeBSD, too.

https://ninja-build.org/manual.html

**Comment by levitte** [comment]
ninja-build has been suggested before, but no one has spent energy on doing it so far.  I did have a look some time ago, but never finished it up.  For *building*, it should be quite simple to create a workable ninja-build template.  I expect it to become just a definition of the common template specific perl functions (those that you see in the block at the end of, for example, `unix-Makefile.tmpl`).  However, for other operations, such as installation, I suspect some other mechanism is required (separate installation script?)...  I have no idea how that kind of thing is done with ninja-based projects.

As for marrying MSVC, mingw with msys and mingw without msys, I think that might be hard, but that might also be my misunderstanding of the components.  In the end, it's not so much about those components as it is about the assume `make` implementation as well as the underlying shell.  `unix-Makefile.tmpl` assumes a POSIX shell and a Unix kind of `make` implementation while `windows-makefile.tmpl` assumes `cmd` and `nmake`.  Those *are* the basic assumptions.  Unfortunately, certain assumptions about the compiler has slipped into `windows-makefile.tmpl` (i.e. it's MSVC or another compiler with compatible options)...  that could certainly be worked on, but you'll still have to battle out the basic assumptions about shell type and `make` implementation.
So the question is how that marriage would work together with those basic assumptions.

**Comment by misery** [comment]
Any chance to allow msys-free build for openssl 3.0.0?

**Comment by misery** [comment]
Ping! :-)
Any opinions?

**Comment by mattcaswell** [comment]
> Any chance to allow msys-free build for openssl 3.0.0?

This is unlikely to receive any 3.0 dev team time (we've got a long list of stuff we need to do and limited time to do it). That's not to say someone external to the 3.0 dev team couldn't contribute something - although it looks like a non-trivial task to me.

**Comment by levitte** [comment]
It's definitely a non-trivial task.  Basically, that would require a GNU-like makefile but with commands for CMD rather than sh.  The way things are currently done, that requires a whole new makefile template.

**Comment by nhorman** [comment]
there is currently a community effort to create either a meson/ninja or cmake build system here:
https://github.com/openssl/openssl/issues/16812
Can this be closed as a duplicate of that?

**Comment by misery** [comment]
If meson/cmake will be used, yes. But is it really accepted now?

**Comment by nhorman** [comment]
No, its not accepted at all, its just in development, but it seems as though for the input above [here](https://github.com/openssl/openssl/issues/6111#issuecomment-385158253), there is no plan to offer the msys changes you are looking for.  It seems the options are 
1) wait for the community developed meson/cmake build system
2) propose the msys changes you would like to see as a community issue

Would you like me to assign this to you as a community issue?

**Comment by nhorman** [comment]
no response, marking as inactive, to be closed at the end of 3.4 dev barring further input
